### ISSUE TITLE
FunctionField doesn't work properly in CustomDatagrid

### ISSUE THREAD

**Issue by niksamoil** [marmelab/react-admin#7544]
**What you were expecting:**
I expected to recieve record in <FunctionField label="Name" render={record => `${record.first_name} ${record.last_name}`} />, which is placed in CustomDatagrid.

This code is from documentation [Datagrid](https://marmelab.com/react-admin/Datagrid.html#body).
My code has almost the same structure, with different fields.
```
// in src/PostList.js
import { Datagrid, DatagridBody, List, TextField } from 'react-admin';
import { TableCell, TableRow, Checkbox } from '@mui/material';

const MyDatagridRow = ({ id, onToggleItem, children, selected, selectable }) => (
    <TableRow key={id}>
        {/* first column: selection checkbox */}
        <TableCell padding="none">
            <Checkbox
                disabled={selectable}
                checked={selected}
                onClick={event => onToggleItem(id, event)}
            />
        </TableCell>
        {/* data columns based on children */}
        {React.Children.map(children, field => (
            <TableCell key={`${id}-${field.props.source}`}>
                {field}
            </TableCell>
        ))}
    </TableRow>
);

const MyDatagridBody = props => <DatagridBody {...props} row={<MyDatagridRow />} />;
const MyDatagrid = props => <Datagrid {...props} body={<MyDatagridBody />} />;

const PostList = () => (
    <List>
        <MyDatagrid>
            <TextField source="title" />
            <FunctionField
                 style={{
                     letterSpacing: '-0.02em',
                     color: '#555555',
                 }}
                label="ra.label.uppercase.supplier"
                render={(record) => `${record.something}`} // record does't work at all
        </MyDatagrid>
    </List>
)

export default PostList;
```

**Other information:**

**Environment**

* React-admin version: 4.0.0
* Last version that did not exhibit the issue (if applicable): 2.9
* React version: 17.0.2
* Browser: Google Chrome Windows

**Comment by fzaninotto** [comment]
Thanks for your feedback! The documentation for custom row elements is missing a key element: each row must instanciate a `RecordContextProvider`. The following should work:

```diff
-import { Datagrid, DatagridBody, List, TextField } from 'react-admin';
+import { Datagrid, DatagridBody, List, TextField, RecordContextProvider } from 'react-admin';
import { TableCell, TableRow, Checkbox } from '@mui/material';

-const MyDatagridRow = ({ id, onToggleItem, children, selected, selectable }) => (
+const MyDatagridRow = ({ id, onToggleItem, children, selected, selectable, record }) => (
+ <RecordContextProvider value={record}>
    <TableRow key={id}>
        {/* first column: selection checkbox */}
        <TableCell padding="none">
            <Checkbox
                disabled={selectable}
                checked={selected}
                onClick={event => onToggleItem(id, event)}
            />
        </TableCell>
        {/* data columns based on children */}
        {React.Children.map(children, field => (
            <TableCell key={`${id}-${field.props.source}`}>
                {field}
            </TableCell>
        ))}
    </TableRow>
+</RecordContextProvider>
);
```

I'm marking this as a documentation problem.

**Comment by niksamoil** [comment]
@fzaninotto thanks, it works. ;)

**Comment by WiXSL** [comment]
Fixed by #7546
